quota/projectquota.go

```diff
@@ -195,23 +195,30 @@ func (q *Control) SetQuota(targetPath string, quota Quota) error {
 		state.Lock()
 		projectID = state.nextProjectID
 
-		//
-		// assign project id to new container directory
-		//
-		err := setProjectID(targetPath, projectID)
-		if err != nil {
+		defer func() {
```

Comment:
I don't understand this change.
The only difference seems to be before we incremented the project ID before setting the quota, now it increments after and I don't see how that would change anything significantly.

Reply:
Not exactly. Previously if we created the same named volume the second time, because the path was still in the map with a project id, it skipped actually setting the project id to the path, which then caused the quota to not actually get set. 

Previously
```go
        projectID, ok := q.quotas[targetPath]
	if !ok {
                // a new path/volume; so need a new project id
		state := getPquotaState()
		projectID = state.nextProjectID

		//
		// assign project id to new container directory
		//
		err := setProjectID(targetPath, projectID)
		if err != nil {
			state.Unlock()
			return err
		}

		state.nextProjectID++

                // save to state
		q.quotas[targetPath] = projectID
	}
        // here means the path/volume was in state
        // but disregards for the case where the volume might've been deleted at some point, 
        // invalidating/freeing the previously set project id.
```

Now, 

```go
	projectID, ok := q.quotas[targetPath]
	if !ok {
                // a new volume; get a new project id
		state := getPquotaState()
		state.Lock()
		projectID = state.nextProjectID

                 // Not safe to increment the project id just yet, in case setProjectId fails
		defer func() {
			if _, exists := q.quotas[targetPath]; exists {
				state.nextProjectID++
			}
			state.Unlock()
		}()
	}


        // Reuse the project id if found in state
        // or the new one from the if block above
	//
	// assign project id to new container directory
	//
	err := setProjectID(targetPath, projectID)
	if err != nil {
		return err
	}

	q.quotas[targetPath] = projectID
```

Please let me know if im missing something.

Reply:
Another bit of background [here](https://github.com/moby/moby/pull/43928?notification_referrer_id=NT_kwDOBIfDWLM0MTUwNTE3Njc4Ojc2MDA2MjMy#discussion_r940613731)

PS: hasn't been repaired yet.